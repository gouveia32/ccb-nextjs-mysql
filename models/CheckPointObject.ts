export interface CheckPointInterface {
    id?: number | string;
    text: string;
    checked: boolean;
  }
  
  export type CheckPointType = CheckPointInterface;
  
  export const CheckPointObject: CheckPointType = {
           id: 0,
           text: "",
           checked: false,
         };
  
  export enum cCheckPoint {
    id = 'id',
    text = 'text',
    checked = 'checked',
  }