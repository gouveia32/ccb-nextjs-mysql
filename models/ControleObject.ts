export interface ControleInterface {
  id?: number | string;
  texto: string;
  marcado: boolean;
}

export type TipoControle = ControleInterface;

export const ControleObject: TipoControle = {
         id: 0,
         texto: "",
         marcado: false,
       };

export enum cControle {
  id = 'id',
  texto = 'texto',
  marcado = 'marcado',
}
