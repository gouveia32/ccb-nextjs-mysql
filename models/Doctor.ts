export interface DoctorModel {
  id: string;
  name: string;
  email: string;
  image?: string;
}

export type DoctorType = DoctorModel;

export const cDoctorModel = {
  id: "id",
  name: "name",
  email: "email",
  image: '',
};

export const DoctorObject: DoctorType = {
  id: "",
  name: "",
  email: "",
  image: '',
};

export function isDoctorType(object: any): object is DoctorType {
  return "name" in object;
}
