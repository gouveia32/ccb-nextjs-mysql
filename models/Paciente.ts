export interface PacienteModel {
  id: string;
  nome: string;
}

export type TipoPaciente = PacienteModel;

export const cPacienteModel = {
  id: "id",
  nome: "nome"
};

export const PacienteObject: TipoPaciente = {
  id: "",
  nome: "",
};

export function isPacienteType(object: any): object is TipoPaciente {
  return "nome" in object;
}
