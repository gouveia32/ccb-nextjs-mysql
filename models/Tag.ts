export interface TagModel {
  id: string;
  name: string;
}

export type TagType = TagModel;

export const cTagModel = {
  id: "id",
  name: "name"
};

export const TagObject: TagType = {
  id: "",
  name: "",
};

export function isTagType(object: any): object is TagType {
  return "name" in object;
}
