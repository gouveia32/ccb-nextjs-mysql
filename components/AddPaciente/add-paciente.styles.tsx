import styled from "styled-components";
import { device } from "../../resources/styles/utils/media-query-utils";
import { motion } from "framer-motion";
import { PacienteTypeEnum } from "../../models/Paciente";

interface AddPacienteInputInterface {
  open: boolean;
  edit?: boolean;
  checkpoints?: number;
  type?: PacienteTypeEnum;
}

export const AddPacienteInput = styled.div<AddPacienteInputInterface>`
  height: ${(props) =>
    props.edit
      ? "max-content"
      : props.open
      ? props.type === PacienteTypeEnum.CHECK &&
        props.checkpoints &&
        props.checkpoints <= 3
        ? `${props.checkpoints * 38 + 162}px`
        : props.type === PacienteTypeEnum.CHECK
        ? "276px"
        : "235px"
      : "45px"};
  width: ${(props) => (props.edit ? "100%" : "500px")};
  padding: 0.6rem;
  border-radius: 0.5rem;
  box-shadow: 0 1px 7px rgb(128 128 128);
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: #fff;

  transition: 0.5s all;

  @media only screen and ${device.mobileL} {
    width: 100%;
  }
`;

export const AddPacienteInputNameInput = styled.input`
  width: 100%;
  border: none;
  outline: none;
  background-color: transparent;
  border-radius: 0.5rem;
  font-weight: bold;

  &:focus {
    border-bottom: 1px solid gray;
  }
`;

export const AddPacienteInputErrorMessage = styled.span`
  position: absolute;
  color: red;
  bottom: 5rem;
`;

export const AddPacienteInputContentWrapper = styled(motion.div)`
  width: 100%;
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const AddPacienteInputContent = styled.textarea`
  margin-top: 1rem;
  width: 100%;
  border: none;
  outline: none;
  border-radius: 0.5rem;
  resize: none;
  background-color: transparent;

  &:focus {
    border-bottom: 1px solid gray;
  }
`;

export const AddPacienteInputTags = styled.div`
  flex: 1;
  display: flex;
  overflow-x: auto;
  padding: 0.3rem 0;

  &::-webkit-scrollbar {
    height: 5px;
  }
`;

export const AddPacienteInputMenu = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
`;

export const AddPacienteInputContentSwitch = styled.div`
  padding: 0 0.3rem;
  border-left: 1px solid lightgray;

  svg {
    cursor: pointer;
    transition: 0.3s transform;
  }
  &:hover {
    svg {
      transform: scale(1.1);
    }
  }
`;

interface AddPacienteInputCheckPointsInterface {
  edit: boolean;
}

export const AddPacienteInputCheckPoints = styled.div<
  AddPacienteInputCheckPointsInterface
>`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;

  max-height: ${(props) => (props.edit ? "300px" : "114px")};
  overflow-y: auto;
`;

export const AddPacienteInputAddCheckPoint = styled.div`
  margin: 0.2rem 0;
  cursor: pointer;

  &:hover {
    svg {
      color: #f5b500;
    }
  }

  svg {
    transition: 0.3s all;
  }
`;

interface AddPacienteInputTagInterface {
  chosen?: boolean;
}

export const AddPacienteInputTag = styled.div<AddPacienteInputTagInterface>`
  margin: 0 0.1rem;
  cursor: pointer;
  border-radius: 1rem;
  padding: 0.2rem 0.5rem;
  white-space: nowrap;

  display: flex;
  align-items: center;

  transition: all 0.3s;

  ${(props) => props.chosen && "background-color: #a5a5a5; color: white;"}

  &:hover {
    background-color: #a5a5a5;
    color: white;
  }
`;
