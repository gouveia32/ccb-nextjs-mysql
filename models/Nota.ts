import { TipoPaciente } from "./Paciente";
import { TipoControle } from "./ControleObject";
import { Nota } from "@prisma/client";

export enum tipoNotaEnum {
  CHECK = "CHECK",
  TEXT = "TEXT",
}

export interface NotaModel {
  id?: string;
  nome: string;
  tipoNota: tipoNotaEnum;
  conteudo: string;
  cor: string;
  pino: boolean;
  criadoEm: string;
  pacientes: TipoPaciente[];
  controles?: TipoControle[];
}

export type TipoNota = NotaModel;

export const NotaObject: TipoNota = {
  id: "",
  nome: "",
  tipoNota: tipoNotaEnum.TEXT,
  conteudo: "",
  cor: "#fff",
  pino: false,
  criadoEm: new Date(Date.now()).toString(),
  pacientes: [],
  controles: [],
};

export const NotaDBObject: Nota = {
  id: "",
  nome: "",
  tipoNota: "TEXT",
  cor: "",
  conteudo: "",
  pino: false,
  criadoEm: new Date(Date.now()),
  medicoId: "",
};

export enum cNotaModel {
  id = "id",
  nome = "nome",
  tipoNota = "tipoNota",
  cor = "cor",
  conteudo = "conteudo",
  pino = "pino",
  criadoEm = "criadoEm",
  pacientes = "pacientes",
  controles = "controles",
}
