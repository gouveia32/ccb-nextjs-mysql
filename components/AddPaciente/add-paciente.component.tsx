import React, { useCallback, useEffect, useRef, useState } from "react";
import { ChangeActionType } from "../../lib/helpers";
import {
  AddPacienteInput,
  AddPacienteInputAddCheckPoint,
  AddPacienteInputCheckPoints,
  AddPacienteInputContent,
  AddPacienteInputContentSwitch,
  AddPacienteInputContentWrapper,
  AddPacienteInputErrorMessage,
  AddPacienteInputMenu,
  AddPacienteInputNameInput,
  AddPacienteInputTag,
  AddPacienteInputTags,
} from "./add-paciente.styles";
import { Button, Divider, IconButton } from "@material-ui/core";
import AddOutlinedIcon from "@material-ui/icons/AddOutlined";
import { SubmitHandler, useForm } from "react-hook-form";
import ColorPicker from "../ColorPicker/color-picker-component";
import ListOutlinedIcon from "@material-ui/icons/ListOutlined";
import TextFieldsOutlinedIcon from "@material-ui/icons/TextFieldsOutlined";
import LabelOutlinedIcon from "@material-ui/icons/LabelOutlined";
import PacienteCheckItem from "./AddPacienteCheckItem/add-paciente-checkitem.component";
import { cPacienteModel, PacienteType, PacienteTypeEnum } from "../../models/Paciente";
import {
  CheckPointObject,
  CheckPointType,
} from "../../models/ControleObject";
import { TagType } from "../../models/Tag";

export interface AddPacienteProps {
  onHandleChange: (action: ChangeActionType) => void;
  pacienteModel: PacienteType;
  tags: TagType[];
  onAddPaciente: () => void;
  onClick?: () => void;
  edit?: boolean;
}

const AddPaciente: React.FC<AddPacienteProps> = ({
  onHandleChange,
  pacienteModel,
  onAddPaciente,
  tags,
  edit,
  onClick,
}: AddPacienteProps) => {
  const [focused, setFocused] = useState<boolean>(!!edit);

  const mainRef = useRef(null);

  const clickOutsideListener = useCallback((e: MouseEvent) => {
    if (!(mainRef.current! as any).contains(e.target)) {
      setFocused(false);
    }
  }, []);

  useEffect(() => {
    document.addEventListener("click", clickOutsideListener);
    return () => {
      document.removeEventListener("click", clickOutsideListener);
    };
  }, [clickOutsideListener]);

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<PacienteType>();

  const onSubmit: SubmitHandler<PacienteType> = (data) => onAddPaciente();

  const handleChangePacienteType = (e: any, pacienteType: PacienteTypeEnum) => {
    onHandleChange({
      attr: cPacienteModel.pacienteType,
      value: pacienteType,
    });
    e.stopPropagation();
  };

  const renderBottomMenu = (
    <AddPacienteInputMenu>
      <AddPacienteInputTags>
        {tags &&
          tags.map((tag: TagType, k: number) => (
            <AddPacienteInputTag
              key={tag.id}
              chosen={!!pacienteModel.tags.find((t) => t.id === tag.id)}
              onClick={() =>
                onHandleChange({ attr: cPacienteModel.tags, value: tag })
              }
            >
              <LabelOutlinedIcon fontSize={"small"} className="me-1" />
              <span>{tag.name}</span>
            </AddPacienteInputTag>
          ))}
      </AddPacienteInputTags>
      <AddPacienteInputContentSwitch>
        <ColorPicker
          onChooseColor={(color) =>
            onHandleChange({ attr: cPacienteModel.color, value: color })
          }
          edit={edit}
        />
      </AddPacienteInputContentSwitch>
      <AddPacienteInputContentSwitch>
        {pacienteModel.pacienteType === PacienteTypeEnum.TEXT ? (
          <IconButton
            size={"small"}
            onClick={(event) => handleChangePacienteType(event, PacienteTypeEnum.CHECK)}
          >
            <ListOutlinedIcon />
          </IconButton>
        ) : (
          <IconButton
            size={"small"}
            onClick={(event) => handleChangePacienteType(event, PacienteTypeEnum.TEXT)}
          >
            <TextFieldsOutlinedIcon />
          </IconButton>
        )}
      </AddPacienteInputContentSwitch>
    </AddPacienteInputMenu>
  );

  const renderNameInput = (
    <>
      <AddPacienteInputNameInput
        placeholder={focused ? "Nome" : "Crie uma nota..."}
        {...register(cPacienteModel.name, { required: true, maxLength: 20 })}
        onChange={(event) =>
          onHandleChange({ attr: cPacienteModel.name, value: event.target.value })
        }
        value={pacienteModel.name}
      />
      {errors.name?.type === "required" && (
        <AddPacienteInputErrorMessage style={{ bottom: 0 }}>
          O Nome é obrigatório
        </AddPacienteInputErrorMessage>
      )}
    </>
  );

  const renderTextContentInput = (
    <>
      <AddPacienteInputContent
        {...register(cPacienteModel.content, { maxLength: 1000 })}
        onFocus={() => setFocused(true)}
        rows={edit ? 8 : 4}
        onChange={(event) =>
          onHandleChange({
            attr: cPacienteModel.content,
            value: event.target.value,
          })
        }
        placeholder={"Escreva algo"}
        value={pacienteModel.content}
      />
      {errors.content?.type === "required" && (
        <AddPacienteInputErrorMessage>Too long</AddPacienteInputErrorMessage>
      )}
    </>
  );

  const renderCheckPointsContentInput = (
    <>
      <AddPacienteInputCheckPoints edit={!!edit}>
        {pacienteModel.checkPoints
          ?.filter((f) => !f.checked)
          .map((c: CheckPointType, i: number) => (
            <PacienteCheckItem
              key={c.id}
              checkItem={c}
              onHandleChange={(newCheckItem) =>
                onHandleChange({
                  attr: cPacienteModel.checkPoints,
                  value: newCheckItem,
                })
              }
              onDelete={(id) =>
                onHandleChange({
                  attr: cPacienteModel.checkPoints,
                  value: id,
                })
              }
            />
          ))}
        {pacienteModel.checkPoints?.find((f) => f.checked) && (
          <Divider className="w-100 bg-dark mt-3 mb-2" />
        )}
        {pacienteModel.checkPoints
          ?.filter((f) => f.checked)
          .map((c: CheckPointType, i: number) => (
            <PacienteCheckItem
              key={c.id}
              checkItem={c}
              onHandleChange={(newCheckItem) =>
                onHandleChange({
                  attr: cPacienteModel.checkPoints,
                  value: newCheckItem,
                })
              }
              onDelete={(id) =>
                onHandleChange({
                  attr: cPacienteModel.checkPoints,
                  value: id,
                })
              }
            />
          ))}
      </AddPacienteInputCheckPoints>
      <AddPacienteInputAddCheckPoint>
        <IconButton
          size={"small"}
          onClick={() =>
            onHandleChange({
              attr: cPacienteModel.checkPoints,
              value: CheckPointObject,
            })
          }
        >
          <AddOutlinedIcon />
        </IconButton>
      </AddPacienteInputAddCheckPoint>
    </>
  );

  return (
    <AddPacienteInput
      edit={edit}
      open={focused}
      checkpoints={pacienteModel.checkPoints?.length}
      type={pacienteModel.pacienteType}
      onFocus={() => setFocused(true)}
      onClick={onClick}
      style={{ backgroundColor: pacienteModel.color }}
      ref={mainRef}
      onKeyDown={(event) =>
        event.code === "Enter" &&
        pacienteModel.pacienteType === PacienteTypeEnum.CHECK &&
        onHandleChange({
          attr: cPacienteModel.checkPoints,
          value: CheckPointObject,
        })
      }
    >
      {renderNameInput}
      {(focused || (edit !== undefined && edit)) && (
        <AddPacienteInputContentWrapper
          initial={{ opacity: 0, width: "100%" }}
          animate={{ opacity: 1 }}
          transition={{ duration: 0.5 }}
        >
          {pacienteModel.pacienteType === PacienteTypeEnum.TEXT
            ? renderTextContentInput
            : renderCheckPointsContentInput}
          {renderBottomMenu}
          <Button
            id={"add-button"}
            fullWidth={true}
            size={"small"}
            variant={"outlined"}
            onClick={handleSubmit(onSubmit)}
            className="mt-2"
            disabled={pacienteModel.name.length === 0}
          >
            {edit ? "Alterar" : "Adicionar"}
          </Button>
        </AddPacienteInputContentWrapper>
      )}
    </AddPacienteInput>
  );
};

export default React.memo(AddPaciente);
