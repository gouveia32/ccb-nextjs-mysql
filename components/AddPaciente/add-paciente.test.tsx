import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import AddPaciente, { AddPacienteProps } from "./add-paciente.component";
import { PacienteObject, PacienteTypeEnum } from "../../models/Paciente";
import { TagObject } from "../../models/Tag";
import {
  AddPacienteInputCheckPoints,
  AddPacienteInputContent,
  AddPacienteInputNameInput,
  AddPacienteInputTag,
} from "./add-paciente.styles";
import { CheckPointObject } from "../../models/ControleObject";
import PacienteCheckItem from "./AddPacienteCheckItem/add-paciente-checkitem.component";
import { Divider } from "@material-ui/core";

describe("AddPaciente component", () => {
  let wrapper: ShallowWrapper;
  let mockOnAddPaciente: any;
  let mockOnHandleChange: any;
  let mockOnClick: any;

  beforeEach(() => {
    mockOnAddPaciente = jest.fn();
    mockOnHandleChange = jest.fn();
    mockOnClick = jest.fn();

    const mockProps: AddPacienteProps = {
      edit: false,
      pacienteModel: PacienteObject,
      tags: [TagObject, TagObject],
      onAddPaciente: mockOnAddPaciente,
      onClick: mockOnClick,
      onHandleChange: mockOnHandleChange,
    };
    wrapper = shallow(<AddPaciente {...mockProps} />);
  });

  it("should render AddPaciente component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call onHandleChange when AddPacienteInputNameInput component is clicked", () => {
    wrapper
      .find(AddPacienteInputNameInput)
      .simulate("change", { target: { value: "" } });

    expect(mockOnHandleChange).toHaveBeenCalled();
  });

  it("should render AddPacienteInputContent component when pacienteType is TEXT and call onHandleChange", () => {
    const pacienteModel = PacienteObject;
    pacienteModel.pacienteType = PacienteTypeEnum.TEXT;
    const mockProps: AddPacienteProps = {
      edit: true,
      pacienteModel: pacienteModel,
      tags: [],
      onAddPaciente: mockOnAddPaciente,
      onClick: mockOnClick,
      onHandleChange: mockOnHandleChange,
    };

    wrapper = shallow(<AddPaciente {...mockProps} />);
    const addPacienteInputContent = wrapper.find(AddPacienteInputContent);
    addPacienteInputContent.simulate("change", { target: { value: "" } });

    expect(addPacienteInputContent).toBeTruthy();
    expect(addPacienteInputContent).toHaveLength(1);
    expect(mockOnHandleChange).toHaveBeenCalled();
    expect(addPacienteInputContent.prop("value")).toBe(mockProps.pacienteModel.content);
  });

  it("should render AddPacienteInputCheckPoints, Divider, PacienteCheckItem when pacienteType is CHECK", () => {
    const pacienteModel = PacienteObject;
    pacienteModel.pacienteType = PacienteTypeEnum.CHECK;

    const checkedCheckpoint = CheckPointObject;
    checkedCheckpoint.checked = true;
    pacienteModel.checkPoints = [CheckPointObject, checkedCheckpoint];

    const mockProps: AddPacienteProps = {
      edit: true,
      pacienteModel: pacienteModel,
      tags: [],
      onAddPaciente: mockOnAddPaciente,
      onClick: mockOnClick,
      onHandleChange: mockOnHandleChange,
    };

    wrapper = shallow(<AddPaciente {...mockProps} />);
    const addPacienteInputCheckPoints = wrapper.find(AddPacienteInputCheckPoints);
    expect(addPacienteInputCheckPoints).toHaveLength(1);

    const pacienteCheckItemChecked: any = wrapper.find(PacienteCheckItem);
    expect(pacienteCheckItemChecked).toHaveLength(2);
    expect(pacienteCheckItemChecked.at(0).prop("checkItem")).toBe(CheckPointObject);
    expect(pacienteCheckItemChecked.at(1).prop("checkItem")).toBe(
      checkedCheckpoint
    );

    const divider = wrapper.find(Divider);
    expect(divider).toHaveLength(1);
  });

  it("should render AddPacienteInputTag components when tags array is not empty", () => {
    const pacienteModel = PacienteObject;
    pacienteModel.pacienteType = PacienteTypeEnum.TEXT;
    const mockProps: AddPacienteProps = {
      edit: true,
      pacienteModel: pacienteModel,
      tags: [TagObject, TagObject],
      onAddPaciente: mockOnAddPaciente,
      onClick: mockOnClick,
      onHandleChange: mockOnHandleChange,
    };

    wrapper = shallow(<AddPaciente {...mockProps} />);
    const addPacienteInputTag = wrapper.find(AddPacienteInputTag);

    expect(addPacienteInputTag).toHaveLength(2);
  });
});
