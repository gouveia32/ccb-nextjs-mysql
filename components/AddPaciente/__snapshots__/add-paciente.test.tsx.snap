// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNote component should render AddNote component 1`] = `ShallowWrapper {}`;
